Add handling of symbolic shapes

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56966 [JIT] Disable Complete Shape Inlining For Testing Purposes
* #56828 Add unary/binary ops necessary for mobilenet
* #56642 Only optimize while change has been made
* #56438 [JIT] integer value refinement
* #56437 Factor out isDominatedBy
* #55978 Add Value * == Value * peephole
* #55977 [JIT] [Easy] Factor out non tensor peephole
* #55926 Add list len refinement
* **#55925 Add handling of symbolic shapes**
* #54809 Initial Symbolic Shape Analysis

This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

Commit:
Add handling of symbolic shapes

[ghstack-poisoned]

Commit:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

[ghstack-poisoned]

Commit:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -87,52 +87,95 @@ struct SymbolicShapeAnalyzer {
+                    use.user->output(), shape[*norm_index].value());
+              } else if (norm_index) {
+                int64_t symbolic_index = shape[*norm_index].value();
+                if (!symbolic_shape_map.count(symbolic_index)) {
+                  symbolic_shape_map[symbolic_index] = {};
+                }
```
Are you sure this is needed? I think accessing a map element via `operator[]` would insert an empty element if it didn't exist (in contrast to `.at`).

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -87,52 +87,95 @@ struct SymbolicShapeAnalyzer {
-  void substituteInputTensorProperties() {
-    for (auto index : tensor_indices) {
-      substituteTensorProperties(index);
+  void substituteInputTensorProperties(bool substitute_symbolic_dims) {
+    std::unordered_map<int64_t, std::vector<Value*>> symbolic_shape_map;
```
Im having hard times following what's going on here: do you mind adding some comments, maybe with examples?

ZolotukhinM replied:
Note that readers of that code would not first see the commit message or the test coming in this PR - it's good for the code to have a similar description too.

Commit:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

Commit:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

Commit:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

eellison commented:
@ZolotukhinM @penguinwu PTAL

penguinwu commented on test/jit/test_symbolic_shape_analysis.py:
```diff
@@ -42,14 +42,34 @@ def foo(x, y):
+        prop_shapes_on_graph([1, 6, 5], [1, 7, 1, 5])
+        FileCheck().check("1, 7, 6, 5").run(foo.graph)
+
+        prop_shapes_on_graph([None, None], [None, None, None])
```
Does `None` mean symbolic shape? But then there is `new_sybolic_shape_symbol` used later. Adding a comment would help.

penguinwu commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -86,54 +103,105 @@ struct SymbolicShapeAnalyzer {
+    // or first dimension, we also try to resolve symbolic shapes of the same
+    // symbolic value to the same Value * in the shape compute graph.
+    // for the shape logic:
+    // dim1 = inp1[0], dim2 = inp2[0]; return dim1 if dim2 == 1 else dim2
```
Break line on the examples to make it easier to read

penguinwu commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -86,54 +103,105 @@ struct SymbolicShapeAnalyzer {
-  void substituteInputTensorProperties() {
-    for (auto index : node_input_tensor_indices) {
-      substituteTensorProperties(index);
+  void substituteInputTensorProperties(bool substitute_symbolic_dims) {
```
What is the difference between `substitute_symbolic_dims` being true or false? Adding a comment would help

penguinwu commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -86,54 +103,105 @@ struct SymbolicShapeAnalyzer {
+            auto index = constant_as<int64_t>(use.user->inputs().at(1));
+            if (index) {
+              auto norm_index = normIndex(*index, *tensor_shape.rank());
+              if (norm_index &&
```
Since both conditions have `norm_index`, it may be more readable to hoist `if (norm_index)` out as a top-level branch

penguinwu commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -86,54 +103,105 @@ struct SymbolicShapeAnalyzer {
-                  use.user->output(), shape[*norm_index].static_size());
-            }
-          }
+  void mergeSymbolicShapeSets(const std::vector<Value*>& symbolic_set) {
```
Do not understand what this tries to do, could you add more comments to explain?

penguinwu reviewed (dismissed):
A couple of places need more explanation for me to understand better. Will do another read after the comments.

Commit:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
Looks good to me, thanks for adding more comments!

eellison committed:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

eellison committed:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750996).

eellison committed:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750996).

eellison committed:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750996).

eellison committed:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750996).

eellison committed:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750996).

eellison committed:
Update on "Add handling of symbolic shapes"


This sets up the initial handling of symbolic shapes. As in the test, it doesn't work perfectly yet because it needs a couple other optimization passes. The basic description is pretty simple: we resolve tensor dimension indices to the same Value *, and before extracting out the output Tensor shape we substitute in symbolic shapes. We don't substitute during optimization because they are represented as negative numbers so we don't want them inadvertently used in Constant prop or something else.

Differential Revision: [D27750996](https://our.internmc.facebook.com/intern/diff/D27750996)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750996).
